File: src/CameraControl.jl
module CameraControl

using Images, ImageShow, Colors

using PythonCall

mutable struct PiGPIOScript
	pig
	index
	pptext
	text

	function PiGPIOScript(pig, text::AbstractString)
		pptext = preprocess_script(text)
		index = pyconvert(Int, pig.store_script(pptext))
		function f(o)
			try
				stop(o)
				delete!(o)
			catch
			end
		end
		finalizer(f, new(pig, index, pptext, text))
	end
	PiGPIOScript(pig, ind::Int) = new(pig, ind, "", "")
end

function Base.show(io::IO, ::MIME"text/plain", scr::PiGPIOScript)
	print(io, "PiGPIOScript($(scr.pig), index=$(scr.index), status=$(status(scr)))")
end

function Base.delete!(scr::PiGPIOScript)
	try
		scr.pig.delete_script(scr.index)
	catch
	end
	scr.index = -1
end

function start(scr::PiGPIOScript, args...)
    scr.pig.run_script(scr.index, collect(args))
end

function stop(scr::PiGPIOScript)
	try
		scr.pig.stop_script(scr.index)
	catch
	end
end

Base.run(scr::PiGPIOScript, args...) = start(scr, args)
halt(scr::PiGPIOScript) = stop(scr)

function status(scr::PiGPIOScript)
    if scr.index < 0
        return :FINALIZED
    end
    e, p = scr.pig.script_status(scr.index)
    e = pyconvert(Int, e)
    if e >= 0
        e = ScriptStatus(e)
    end
    return e
end

function params(scr::PiGPIOScript)
    if scr.index < 0
        return nothing
    end
    e, p = scr.pig.script_status(scr.index)
    e = pyconvert(Int, e)
    if e >= 0
        e = ScriptStatus(e)
        p = Int64.(reinterpret(UInt32, Int32.(collect(pyconvert(Tuple, p)))))
    else
        p = nothing
    end
    return p
end

Base.delete!(scr::Int) = Base.delete!(PiGPIOScript(pig[], scr))
Base.run(scr::Int) = Base.run(PiGPIOScript(pig[], scr))

start(scr::Int) = start(PiGPIOScript(pig[], scr))
stop(scr::Int) = stop(PiGPIOScript(pig[], scr))
halt(scr::Int) = halt(PiGPIOScript(pig[], scr))
status(scr::Int) = status(PiGPIOScript(pig[], scr))
params(scr::Int) = params(PiGPIOScript(pig[], scr))

export PiGPIOScript, stop, start, halt, status, params

const v4l2py = Ref{Py}()
const pigpio = Ref{Py}()
const pig = Ref{Py}()

function __init__()
	v4l2py[] = pyimport("v4l2py")
	pigpio[] = pyimport("pigpio")
end

function start_pigpio(host="localhost", port=8888)
	pig[] = pigpio[].pi(host, port)
	if ! pyconvert(Bool, pig[].connected)
		error("Couldn't open connection to pigpiod")
	end
end
export start_pigpio

function start_pig(host="localhost", port=8888)
    pig = pigpio[].pi(host, port)
    if !pyconvert(Bool, pig.connected)
        error("Couldn't open connection to pigpiod")
    end
	return pig
end
export start_pig

@enum ScriptStatus begin
	INITING = 0
	HALTED = 1
	RUNNING = 2
	WAITING = 3
	FAILED = 4
end
export ScriptStatus
export INITING, HALTED, RUNNING, WAITING, FAILED

function preprocess_script(scr)
	# split the script into lines
	lines = split(scr, "\n")
	# split each line at the first comment character, keeping the leading text
	uncommented = (line->first(split(line, "#"))).(lines)
	# strip leading and trailing whitespace from each line
	stripped = strip.(uncommented)
	# keep only nonempty lines
	nonempty = filter(line->!isempty(line), stripped)
	# rejoin the nonempty lines into a single string
	return join(nonempty, "\n")
end

export trigger_wave_script
function trigger_wave_script(pig;
	TRIG_IN = 25,
	TRIG_TIME = 0,
	TRIG_OUT = 5,
	TRIG_WIDTH = 50,
	LED_IN = 22,
	LED_TIME = 5555,
	LED_OUT = 17,	
	LED_WIDTH = 250,
	STROBE_IN = 6,
	WAVE_DURATION = 33400,
	G1 = 23,
	G2 = 27
)
	p = pigpio[]
	
	for pin ∈ [G1, G2, TRIG_OUT, LED_OUT]
		pig.set_mode(pin, p.OUTPUT)
	end
	for pin ∈ [TRIG_IN, LED_IN, STROBE_IN]
		pig.set_mode(pin, p.INPUT)
	end

	dtled = max(0, LED_TIME - TRIG_WIDTH - TRIG_TIME)
	dtif = max(0, WAVE_DURATION - LED_TIME - LED_WIDTH - TRIG_TIME - TRIG_WIDTH)

	wave = @py []
	wave.append(p.pulse(0, 0, TRIG_TIME))
	wave.append(p.pulse(0, 1<<TRIG_OUT, TRIG_WIDTH))
	wave.append(p.pulse(1<<TRIG_OUT, 0, dtled))
	wave.append(p.pulse(0, 1<<LED_OUT, LED_WIDTH))
	wave.append(p.pulse(1<<LED_OUT, 0, dtif))
	pig.wave_add_generic(wave)
	wave_id = pyconvert(Int, pig.wave_create())

	script = PiGPIOScript(pig, """
	pads 0 16							# set pad drivers to 16 mA
	lda 0 sta p0

	lda $TRIG_IN sta p1
	lda $TRIG_TIME sta p2
	lda $LED_TIME sta p3
	lda $LED_WIDTH sta p4

tag 100
	tick sta p5
	w $G1 1
tag 101	mics 1	r $TRIG_IN	jz 101				# wait for TRIG_IN to go high
	tick sub p5 sta p6
tag 102	inr p0	mics 1 r $TRIG_IN	jnz 102		# wait for falling edge on TRIG_IN
	tick sub p5 sta p7
	w $G1 0

	wvtx $wave_id 						# trigger the wave created above

	w $G2 1
# tag 104	mics 1	r $STROBE_IN	jz 104	# wait for STROBE_IN to go high
# tag 105	mics 1	r $STROBE_IN	jnz 105	# wait for STROBE_IN to go low
# tag 106	mics 1	r $STROBE_IN	jz 106	# wait for STROBE_IN to go high
# tag 107	mics 1	r $STROBE_IN	jnz 107	# wait for STROBE_IN to go low
	tick sub p5 sta p8
tag 103	wvbsy	jnz 103 				# wait for wave to finish
	tick sub p5 sta p9
	# wvdel $wave_id 					# release the wave resources
	w $G2 0

	ret
	"""
    )
end

function trigger_loop(pig; n=100, t_min=1200, t_max=1200+8333, kwargs...)
	c = (t_max-t_min) ÷ n
	time = @elapsed for i = 0:n
		s = trigger_wave_script(pig; LED_TIME=t_min+i*c, kwargs...)
		while status(s)==INITING; end
		start(s)
		# while status(s)!=RUNNING;  sleep(0.001); end
		while status(s)==RUNNING; end
		delete!(s)
		pig.wave_clear()
		# finalize(s)
	end
	@info "$(n/time) fps"
end
export trigger_loop



end # module CameraControl

File: python/CameraControl.py
from enum import Enum
# import v4l2py
import pigpio

# from PIL import Image, ImageShow, ImageColor

class ScriptStatus(Enum):
	INITING = 0
	HALTED = 1
	RUNNING = 2
	WAITING = 3
	FAILED = 4

def preprocess_script(text):
	# Split the script into lines
	lines = text.split('\n')
	# Split each line at the first comment character, keeping the leading text
	uncommented = [line.split('#')[0] for line in lines]
	# Strip leading and trailing whitespace from each line
	stripped = [line.strip() for line in uncommented]
	# Keep only nonempty lines
	nonempty = [line for line in stripped if line]
	# Rejoin the nonempty lines into a single string
	return '\n'.join(nonempty)

class PiGPIOScript:
	def __init__(self, pig, text=None, id=None):
		if text is not None:
			self.pptext = preprocess_script(text)
			self.id = pig.store_script(self.pptext)
		elif id is not None:
			self.id = id
			self.pptext = ''
			self.text = ''
		self.pig = pig
		self.text = text if text else ''
		self.waves = []
		# print(f"{self} created")

	def __del__(self):
		# print(f"{self} finalizing")
		self.stop()
		self.delete()

	def __str__(self):
		return f"PiGPIOScript({self.pig}, id={self.id})"
	
	def __repr__(self):
		return self.__str__()

	def delete(self):
		if self.id >= 0:
			self.pig.delete_script(self.id)
			self.id = -1
	
	def start(self, *args):
		if self.id >= 0:
			self.pig.run_script(self.id, list(args))
	
	def run(self, *args):
		if self.id >= 0:
			self.pig.run_script(self.id, list(args))
	
	def stop(self):
		if self.id >= 0:
			self.pig.stop_script(self.id)

	def status(self):
		e, _ = self.pig.script_status(self.id)
		return ScriptStatus(e)

	def params(self):
		_, p = self.pig.script_status(self.id)
		return p

	def set_params(self, *args):
		if self.id >= 0:
			self.pig.update_script(self.id, list(args))

	def params_valid(self):
		params = self.params()
		hi_to_lo = params[8] - params[6]
		return (0 < hi_to_lo < 8340)

	def param_summary(self):
		params = self.params()
		start_gpio = f"{params[4]:08x}"
		start_to_hi = params[6] - params[5]
		triglo_gpio = f"{params[7]:08x}"
		hi_to_lo = params[8] - params[6]
		lo_to_finish = params[9] - params[8]
		
		summary = f"{start_gpio} {start_to_hi:5d} TrigHI {hi_to_lo:5d} {triglo_gpio} {lo_to_finish:5d} Finish"
		return summary

	def halted(self):
		return self.status() == ScriptStatus.HALTED

	def initing(self):
		return self.status() == ScriptStatus.INITING

	def running(self):
		return self.status() == ScriptStatus.RUNNING


def start_pig(host="localhost", port=8888):
	pig = pigpio.pi(host, port)
	if not pig.connected:
		raise ValueError("Couldn't open connection to pigpiod")
		# raise Exception('Could not connect to pigpio')
	return pig

class PiGPIOWave:
	def __init__(self, pig, config):
		self.pig = pig
		self.config = config
		self.kwargs = config.__dict__
		self.wave = []
		self.id = self.generate_wave()

	def __str__(self):
		return f"PiGPIOWave{self.pig}, id={self.id})"
	
	def __repr__(self):
		return self.__str__()

	def __del__(self):
		self.delete()

	def delete(self):
		if self.id >= 0:
			self.pig.wave_delete(self.id)
			self.id = -1

	def generate_wave(self):
		cf = self.config

		for pin in [cf.TRIG_OUT, cf.RED_OUT, cf.GRN_OUT, cf.BLU_OUT, cf.LED_OUT]:
			self.pig.set_mode(pin, pigpio.OUTPUT)

		for pin in [cf.TRIG_IN, cf.LED_IN, cf.STROBE_IN]:
			self.pig.set_mode(pin, pigpio.INPUT)

		dtled = max(0, cf.LED_TIME - cf.TRIG_WIDTH - cf.TRIG_TIME)
		dtif = max(0, cf.WAVE_DURATION - cf.LED_TIME - cf.LED_WIDTH - cf.TRIG_TIME - cf.TRIG_WIDTH)

		LED_ON_HIGH = cf.LED_MASK
		LED_ON_LOW = 0
		LED_OFF_HIGH = 0
		LED_OFF_LOW = cf.LED_MASK

		self.wave.append(pigpio.pulse(0, 0, cf.TRIG_TIME))
		self.wave.append(pigpio.pulse(0, 1<<cf.TRIG_OUT, cf.TRIG_WIDTH))
		self.wave.append(pigpio.pulse(1<<cf.TRIG_OUT, 0, dtled))
		self.wave.append(pigpio.pulse(LED_ON_HIGH, LED_ON_LOW, cf.LED_WIDTH))
		self.wave.append(pigpio.pulse(LED_OFF_HIGH, LED_OFF_LOW, dtif))
		self.pig.wave_add_generic(self.wave)
		return self.pig.wave_create()


def trigger_wave_script(pig, config):
	script = f"""
	pads 0 16								# set pad drivers to 16 mA

	# we expect wave id in p0
	lda {config.TRIG_IN} sta p1
	lda {config.STROBE_IN} sta p2

	ld p3 1				# status: starting

tag 100
	lda p0         		# load the current value of p0
	or 0 jp 101    		# if p0 is valid (>=0), proceed
	ld p3 0				# status: sequencing paused
	mils 1      		# otherwise delay for a bit
	jmp 100       		# and try again

# 	ld p3 2 			# status: waiting for p2 low
# tag 110
# 	r p2 jnz 110		# wait if p2 is high
# 	ld p3 3				# status: waiting for p2 high
# tag 111
# 	r p2 jz 111			# wait for rising edge of p2

	ld p3 4				# status: sequencing started
	br1 sta v4			# capture starting GPIO in p4
	tick sta v5			# capture the start time in p5

tag 101
	r p1 jz 101			# wait for p1 to go high
	tick sta v6			# capture trigger high time in p6

	ld p3 5
tag 102
	r p1 jnz 102		# wait for falling edge on p1
	br1 sta v7			# capture GPIO at trigger low in p7
	tick sta v8			# capture trigger low time in p8
	ld p3 6				# status: wave tx started
	wvtx p0				# trigger the wave

tag 103
	wvbsy
	jnz 103 			# wait for wave to finish
	tick sta v9			# capture wave finish time in p9
	ld p3 7				# status: wave tx complete
	ld p4 v4 ld p5 v5 ld p6 v6 ld p7 v7 ld p8 v8 ld p9 v9
	ld p3 8				# status: return params updated 

	jmp 100				# do it again
	ret
	"""

	return PiGPIOScript(pig, script)

import time

def trigger_loop(pig, n=1000, t_min=2777, t_max=2778+8333, **kwargs):
	c = int((t_max - t_min) // n)

	start_time = time.time()

	for i in range(n + 1):
		s = trigger_wave_script(pig, LED_TIME=t_min + i * c, **kwargs)
		
		while s.initing():
			pass
		
		s.start()
		
		while s.running():
			pass
		
		p = s.params()
		print_summary(p)

		s.delete()
		pig.wave_clear()

	# capture the end time
	end_time = time.time()

	# Calculate the elapsed time
	elapsed_time = end_time - start_time
	
	print(f"{n/elapsed_time} fps")
	
	return fps



if __name__ == "__main__":
	pig = start_pig()
	trigger_loop(pig)

